forwardproxy.go

```diff
@@ -395,21 +398,58 @@ func (fp *ForwardProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int,
 			r.Header.Add("Via", strconv.Itoa(r.ProtoMajor)+"."+strconv.Itoa(r.ProtoMinor)+" caddy")
 		}
 
-		if fp.responseTimeout != nil {
-			targetConn.SetDeadline(time.Now().Add(*fp.responseTimeout))
-		}
-
+		var err error
 		var response *http.Response
-		err = r.Write(targetConn)
-		if err != nil {
-			return http.StatusBadGateway, errors.New("failed to write http request: " + err.Error())
+		if fp.upstream == nil {
+			// non-upstream request uses httpTransport to reuse connections
+			if r.Body != nil &&
+				(r.Method == "GET" || r.Method == "HEAD" || r.Method == "OPTIONS" || r.Method == " TRACE") {
+				// make sure request is idempotent and could be retried by saving the Body
```

Comment:
None of these methods is allowed to have a body, so this comment is confusing.  I think it's fine to copy the body, but the comment should probably be clear that the body is normally expected to be empty.

Reply:
Added comment